plugins/common/cookie/cookie.go

```diff
@@ -0,0 +1,89 @@
+package cookie
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+	"github.com/sirupsen/logrus"
+)
+
+type CookieAuthConfig struct {
+	URL    string `toml:"cookie_auth_url"`
+	Method string `toml:"cookie_auth_method"`
+
+	// HTTP Basic Auth Credentials
+	Username string `toml:"cookie_auth_username"`
+	Password string `toml:"cookie_auth_password"`
+
+	Body    string          `toml:"cookie_auth_body"`
+	Renewal config.Duration `toml:"cookie_auth_renewal"`
+
+	client *http.Client
+}
+
+func (c *CookieAuthConfig) Start(client *http.Client) (err error) {
+	c.client = client
+
+	if c.Method == "" {
+		c.Method = http.MethodPost
+	}
+
+	// add cookie jar to HTTP client
+	if c.client.Jar, err = cookiejar.New(nil); err != nil {
+		return err
+	}
+
+	// continual auth renewal if set
+	if c.Renewal > 0 {
+		ticker := time.NewTicker(time.Duration(c.Renewal))
+		go func() {
+			for range ticker.C {
+				if err := c.auth(); err != nil {
+					logrus.WithError(err).Error("cookie auth renewal failure")
+				}
```

Comment:
How about using a channel like (untested):
```suggestion
				if err := c.auth(); err != nil {
					select {
						case c.ErrorChan <- err:
						default:
					}
				}
```
You need to init `c.ErrorChan = make(chan error)` in `Start()` then and the caller can the read from the channel to log the errors. We could even provide a convenience function for that passing in a `telegraf.Logger` instance...

Reply:
`error` channel or embedded `Log` is good... only issue is since this package is in `common` and used in another `common` pkg (`httpconfig`), I'm not sure how the original caller would use either solution.

I can refactor `httpconfig` to embed a `Log` field (and pass it into the CookieAuth), but I'm not sure how `Log` would even get set on input/output plugins e.g. in `plugins/inputs/http/http.go`.

> Almost. I had one idea about passing the `auth()` errors to the caller using a non-blocking channel. Maybe you can check if this is doable.

My intention with returning the `auth()` error in a blocking fashion was to raise an immediate failure in the case of bad creds in a config. 
I figured if the initial auth() worked (with good creds), then future auth failures would be intermittent (and non blocking.)

Reply:
Well you could do the first call to `auth()` separately in the function before entering the timer and return the error code in a blocking fashion.
However, for the further calls to auth I'd like to have a way for the caller to know if everything is alright still. You could also provide a callback function. No matter what you decide to do, the caller should have a way to react to the failing auth. 
Maybe the best way would be to just provide an exported `Auth()` function and let the caller do the timing... :-)